perfkitbenchmarker/linux_virtual_machine.py

```diff
@@ -954,3 +956,191 @@ def MoveFile(self, target, source_path, remote_path=''):
+        while True:
+            ready = True
+            status = yaml.load(self.juju_status())
+            # logging.warn(status)
+            for service in status['services']:
+                ss = status['services'][service]['service-status']['current']
+
+                # Accept blocked because the service may be waiting on relation
+                if ss not in ['active', 'unknown']:
+                    ready = False
+
+                if ss in ['error']:
+                    # The service has failed to deploy.
+                    debuglog = self.juju_run('juju debug-log --limit 200')
+                    logging.warn(debuglog)
+                    return False
+
+                for unit in status['services'][service]['units']:
+                    unit_data = status['services'][service]['units'][unit]
+                    ag = unit_data['agent-state']
+                    if ready and ag != 'started':
+                        ready = False
+
+                    ws = unit_data['workload-status']['current']
+                    if ready and ws not in ['active', 'unknown']:
+                        ready = False
+            # TODO: Exit loop if timeout exceeded
+            if ready:
+                return
+            time.sleep(30)
+
+
+    def Deploy(self, charm, units=1):
+        if self.isController:
+            # Check if the charm is already deployed?
+            resp, _ = self.RemoteHostCommand(
+                "juju deploy %s --num-units=%d" % (charm, units))
+
+            return True
+        pass
+
+    def Relate(self, a, b):
+        resp, _ = self.RemoteHostCommand("juju add-relation %s %s" % (a, b))
+        pass
+
+    def AuthenticateVm(self):
+        super(JujuMixin, self).AuthenticateVm()
+
+    def Install(self, package_name):
+        """Installs a PerfKit package on the VM."""
+        if package_name not in self._installed_packages:
+            package = linux_packages.PACKAGES[package_name]
+            package.JujuInstall(self)
+            self._installed_packages.add(package_name)
+
+    def SetupPackageManager(self):
+        if self.isController:
+            resp, _ = self.RemoteHostCommand(
+                "sudo add-apt-repository ppa:juju/stable"
+            )
+        super(JujuMixin, self).SetupPackageManager()
+
+    def PrepareVMEnvironment(self):
+        super(JujuMixin, self).PrepareVMEnvironment()
+        if self.isController:
+            try:
+                self.InstallPackages('juju')
+
+                self.juju_configure_environment()
+
+                self.AuthenticateVm()
+
+                self.juju_bootstrap()
+
+                # Install the Juju agent on the other VMs
+                for unit in self.units:
+                    unit.controller = self
+                    self.juju_addmachine(unit.internal_ip)
+
+            except errors.VirtualMachine.RemoteCommandError as e:
+                raise e
+
+    @vm_util.Retry(log_errors=False, poll_interval=1)
```

Comment:
Unless I'm missing something, it doesn't look like you need this method defined here